File: sohee-jung/week3/백준_gold_불!.java
import java.io.*;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int R, C;
    static char[][] maze;
    static Queue<Pair> fire = new LinkedList<>();
    static Queue<Pair> jihoon = new LinkedList<>();

    static int[] dx = { 1, -1, 0, 0 };
    static int[] dy = { 0, 0, 1, -1 };

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        maze = new char[R][C];

        for (int r = 0; r < R; r++) {
            String line = br.readLine();
            for (int c = 0; c < C ; c++) {
                maze[r][c] = line.charAt(c);

                if (maze[r][c] == 'J') {
                    jihoon.add(new Pair(r, c));
                } else if (maze[r][c] == 'F') {
                    fire.add(new Pair(r, c));
                }
            }
        }

        int result = bfs();
        if (result == -1){
            bw.write("IMPOSSIBLE\n");
        } else {
            bw.write(result + "\n");
        }

        bw.flush();
        bw.close();
        br.close();
    }

    private static int bfs(){
        int time = 0;

        while (!jihoon.isEmpty()) {
            time++;

            int fireSize = fire.size();
            while (fireSize-- > 0) {
                Pair f = fire.poll();
                for (int i = 0; i < 4; i++) {
                    int nx = f.x + dx[i];
                    int ny = f.y + dy[i];
                    if (nx >= 0 && ny >= 0 && nx < R && ny < C) {
                        if (maze[nx][ny] == '.' || maze[nx][ny] == 'J') {
                            maze[nx][ny] = 'F';
                            fire.add(new Pair(nx, ny));
                        }
                    }
                }
            }

            int jihoonSize = jihoon.size();
            while (jihoonSize-- > 0) {
                Pair j = jihoon.poll();
                for (int i = 0; i < 4; i++) {
                    int nx = j.x + dx[i];
                    int ny = j.y + dy[i];
                    if (nx < 0 || ny < 0 || nx >= R || ny >= C) {
                        return time;
                    }
                    if (maze[nx][ny] == '.') {
                        maze[nx][ny] = 'J';
                        jihoon.add(new Pair(nx, ny));
                    }
                }
            }
        }
        return -1;
    }

}

class Pair {
    int x, y;
    Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

File: minseo-song/week3/백준_gold_불!.kt
package week3

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.LinkedList
import java.util.Queue
import java.util.StringTokenizer

data class Point(val x: Int, val y: Int)

fun main() {
    val bufferedReader = BufferedReader(InputStreamReader(System.`in`))
    val bufferedWriter = BufferedWriter(OutputStreamWriter(System.out))

    val st = StringTokenizer(bufferedReader.readLine())
    val r = st.nextToken().toInt()
    val c = st.nextToken().toInt()

    val fireTime = Array(r) { IntArray(c) { -1 } }
    val jihunTime = Array(r) { IntArray(c) { -1 } }

    val fireQueue: Queue<Point> = LinkedList()
    val jihunQueue: Queue<Point> = LinkedList()

    val map = Array(r) { CharArray(c) }

    for (i in 0 until r){
        val line = bufferedReader.readLine()
        for (j in 0 until c){
            map[i][j] = line[j]
            when(map[i][j]){
                'F' -> {
                    fireQueue.add(Point(i, j))
                    fireTime[i][j] = 0
                }
                'J' -> {
                    jihunQueue.add(Point(i, j))
                    jihunTime[i][j] = 0
                }
            }
        }
    }

    val dx = intArrayOf(-1, 1, 0 ,0)
    val dy = intArrayOf(0, 0, -1, 1)

    while (fireQueue.isNotEmpty()) {
        val (x, y) = fireQueue.poll()
        for (i in 0 until 4) {
            val nx = x + dx[i]
            val ny = y + dy[i]
            if (nx in 0 until r && ny in 0 until c) {
                if (map[nx][ny] != '#' && fireTime[nx][ny] == -1) {
                    fireTime[nx][ny] = fireTime[x][y] + 1
                    fireQueue.add(Point(nx, ny))
                }
            }
        }
    }

    while (jihunQueue.isNotEmpty()) {
        val (x, y) = jihunQueue.poll()
        if (x == 0 || x == r - 1 || y == 0 || y == c - 1) {
            bufferedWriter.write("${jihunTime[x][y] + 1}")
            bufferedWriter.close()
            return
        }

        for (i in 0 until 4) {
            val nx = x + dx[i]
            val ny = y + dy[i]
            if (nx in 0 until r && ny in 0 until c) {
                if (map[nx][ny] == '.' && jihunTime[nx][ny] == -1) {
                    if (fireTime[nx][ny] == -1 || jihunTime[x][y] + 1 < fireTime[nx][ny]) {
                        jihunTime[nx][ny] = jihunTime[x][y] + 1
                        jihunQueue.add(Point(nx, ny))
                    }
                }
            }
        }
    }

    bufferedWriter.write("IMPOSSIBLE")
    bufferedWriter.close()
}